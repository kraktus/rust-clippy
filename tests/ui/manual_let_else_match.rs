#![allow(unused_braces, unused_variables, dead_code)]
#![allow(clippy::collapsible_else_if, clippy::let_unit_value)]
#![warn(clippy::manual_let_else)]
// Ensure that we don't conflict with match -> if let lints
#![warn(clippy::single_match_else, clippy::single_match)]

enum Variant {
    Foo,
    Bar(u32),
    Baz(u32),
}

fn f() -> Result<u32, u32> {
    Ok(0)
}

fn g() -> Option<()> {
    None
}

fn h() -> Variant {
    Variant::Foo
}

fn main() {}

fn fire() {
    let v = match g() {
        Some(v_some) => v_some,
        None => return,
    };

    let v = match g() {
        Some(v_some) => v_some,
        _ => return,
    };

    loop {
        // More complex pattern for the identity arm
        let v = match h() {
            Variant::Foo => continue,
            Variant::Bar(v) | Variant::Baz(v) => v,
        };
    }

    // There is a _ in the diverging arm
    // TODO also support unused bindings aka _v
    let v = match f() {
        Ok(v) => v,
        Err(_) => return,
    };
}

fn not_fire() {
    // Multiple diverging arms
    let v = match h() {
        Variant::Foo => panic!(),
        Variant::Bar(_v) => return,
        Variant::Baz(v) => v,
    };

    // Multiple identity arms
    let v = match h() {
        Variant::Foo => panic!(),
        Variant::Bar(v) => v,
        Variant::Baz(v) => v,
    };

    // No diverging arm at all, only identity arms.
    // This is no case for let else, but destructuring assignment.
    let v = match f() {
        Ok(v) => v,
        Err(e) => e,
    };

    // The identity arm has a guard
    let v = match h() {
        Variant::Bar(v) if g().is_none() => v,
        _ => return,
    };

    // The diverging arm has a guard
    let v = match f() {
        Err(v) if v > 0 => panic!(),
        Ok(v) | Err(v) => v,
    };

    // The diverging arm creates a binding
    let v = match f() {
        Ok(v) => v,
        Err(e) => panic!("error: {e}"),
    };
}
