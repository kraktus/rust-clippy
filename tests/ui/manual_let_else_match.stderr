error: this could be rewritten as `let else`
  --> $DIR/manual_let_else_match.rs:28:5
   |
LL | /     let v = match g() {
LL | |         Some(v_some) => v_some,
LL | |         None => return,
LL | |     };
   | |______^
   |
   = note: `-D clippy::manual-let-else` implied by `-D warnings`

error: this could be rewritten as `let else`
  --> $DIR/manual_let_else_match.rs:33:5
   |
LL | /     let v = match g() {
LL | |         Some(v_some) => v_some,
LL | |         _ => return,
LL | |     };
   | |______^

error: this could be rewritten as `let else`
  --> $DIR/manual_let_else_match.rs:40:9
   |
LL | /         let v = match h() {
LL | |             Variant::Foo => continue,
LL | |             Variant::Bar(v) | Variant::Baz(v) => v,
LL | |         };
   | |__________^

error: this could be rewritten as `let else`
  --> $DIR/manual_let_else_match.rs:48:5
   |
LL | /     let v = match f() {
LL | |         Ok(v) => v,
LL | |         Err(_) => return,
LL | |     };
   | |______^

error: aborting due to 4 previous errors

